This improved regex parses all of our network and system logs properly

johann8384 opened the pull request:
Some lines that were dropped:

<27>1 2015-03-02T22:53:45-08:00 localhost.localdomain puppet-agent 5334 -     mirrorurls.extend(list(self.metalink_data.urls()))
<29>1 2015-03-03T06:49:08-08:00 localhost.localdomain puppet-agent 51564 - (/Stage[main]/Users_prd/Ssh_authorized_key[1063-username]) Dependency Group[group] has failures: true
<30>1 2015-03-03T06:49:08-08:00 localhost.localdomain /usr/sbin/gmond 38163 - Error 1 sending the modular data for proc_run#012
<30>1 2015-03-03T06:47:18-08:00 localhost.localdomain /usr/sbin/gmond 31411 - Error 1 sending the modular data for heartbeat#012
<86>1 2015-03-02T22:23:07-08:00 localhost.localdomain sshd 17992 - Accepted publickey for username from 127.0.0.1 port 44656 ssh2
<142>1 2015-03-02T22:23:07-08:00 localhost.localdomain Keepalived_vrrp 21125 - VRRP_Instance(VI_1) ignoring received advertisment...
<142>1 2015-03-02T22:23:07-08:00 localhost.localdomain Keepalived_vrrp 17875 - bogus VRRP packet received on bond0 !!!
<158>1 2015-03-02T22:23:07-08:00 localhost.localdomain fpc0  - PFE_FW_SYSLOG_IP: FW: ae3.0        D  tcp 192.168.1.2 192.168.1.1 47320    25 (1 packets)

johann8384 committed:
This improved regex parses all of our network and system logs properly

otoolep commented:
Thanks @johann8384 -- to be clear, you configured your syslog sender to it sent messages that adhered to the template described in the README?

otoolep commented:
Github formatting may have messed up the lines you supplied, so if you add them to the actual source code -- in the tests -- I can see exactly what was dropped.

johann8384 commented:
Yes, Syslog is one of those specifications that vendors really like to take
their liberty with. So we get logs from all kinds of devices and they don't
conform so strictly. Passing them through syslog-ng (with that template)
and then to gollector results in most of them passing. I suspect it can't
parse all of them and just passes them through as best it can.

With this change gollector can collect from everything directly and only 1%
or so drops. At 20k lines per second, that is pretty good. The stuff that
still drops is pretty badly formatted.

The regex, template,  and RFC stuff is nice, but in the real world you
aren't going to have such conformity. Just try writing a SIP proxy next. :)
On Mar 3, 2015 10:38 AM, "otoolep" <notifications@github.com> wrote:

> Github formatting may have messed up the lines you supplied, so if you add
> them to the actual source code -- in the tests -- I can see exactly what
> was dropped.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/otoolep/syslog-gollector/pull/19#issuecomment-76981707>
> .
>

otoolep commented:
Thanks @johann8384. I agree conformity can be difficult to achieve, and that much of the real world is messier.

The idea behind the template is a) it can be added to existing rsyslog and syslog-ng systems without affecting existing streams and b) I'd rather have guaranteed parsing if you do follow the template, versus trying to parse a whole universe of log line formats, which would be much more difficult. Perhaps if parsing is enabled, gollector should still pass the messages through anyway? It could be a flag.

Let me test your changes, and if they are a super-set of the parsing that is there now, I'll merge.

otoolep commented:
I added a random 3 of the lines your supplied above to the unit test and they parse without change to the regex:

https://github.com/otoolep/syslog-gollector/commit/85f2703c6f6a46854f8239163869e7939c2663d3

So it's not clear to me now why we should merge the regex change. Can you provide me with a line that actually fails parsing?

johann8384 commented:
I really don't care that  much, I submitted the regex because it's an
improvement on the regex that was there. I added some logging for dropped
lines then started tweaking the regex until the dropped lines were
accepted. Then tested previous lines to see they were still accepted.

I played with this one evening, I'm not actually using it. Because I played
with it and made what seems like an improvement, I wanted to share it with
you.

This was the final regex that seemed to have nearly 100% parsing success
against our log stream:

(?s)<([0-9]{1,3})>([0-9])\s(\d{4}-\d{2}-\d{2}T\d{2}\:\d{2}\:\d{2}[\S0-9]*[+-]\d{2}\:\d{2})\s+([\S]+)\s([0-9\S]+)\s+([\S]*)\s+-\s+([A-Za-z0-9_]+)?:?(.*$)

-Jonathan


On Wed, Mar 4, 2015 at 11:47 AM, otoolep <notifications@github.com> wrote:

> I added a random 3 of the lines your supplied above to the unit test and
> they parse without change to the regex:
>
> 85f2703
> <https://github.com/otoolep/syslog-gollector/commit/85f2703c6f6a46854f8239163869e7939c2663d3>
>
> So it's not clear to me now why we should merge the regex change. Can you
> provide me with a line that actually fails parsing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/otoolep/syslog-gollector/pull/19#issuecomment-77231861>
> .
>

otoolep commented:
Thanks @johann8384, I appreciate the patch. Without samples of lines that actually fail though, I can't actually verify it's an improvement.

otoolep commented:
This regex fails to parse RFC5424 timestamps in the form 2003-08-24T05:14:15.000003-07:00.

otoolep commented:
I actually need to work on the regex further, since the currently committed version doesn't parse all the required timestamps either.

Thanks for the suggestion @johann8384
